File: app/src/main/java/com/mkrworld/libfilter/filter/BaseFilter.java
package com.mkrworld.libfilter.filter;

public abstract class BaseFilter {

    private int[] mPixelArray;
    private int mImageWidth;

    /**
     * Constructor
     *
     * @param pixelArray Array of Pixel
     * @param imageWidth
     */
    public BaseFilter(int[] pixelArray, int imageWidth) {
        mPixelArray = pixelArray;
        mImageWidth = imageWidth;
    }

    /**
     * Method to apply Filter and Get final Value
     *
     * @return Filtered PixelArray
     */
    public abstract int[] applyFilter() throws Exception;

    /**
     * Method to get the Image Width
     *
     * @return
     */
    protected int getImageWidth() {
        return mImageWidth;
    }

    /**
     * Method to get the Image Pixel Array
     *
     * @return
     */
    protected int[] getPixelArray() {
        return mPixelArray;
    }
}

File: app/src/main/java/com/mkrworld/libfilter/filter/BaseFilter.kt
package com.mkrworld.libfilter.filter

import com.mkrworld.libfilter.dto.FilterMatrix
import java.util.*

/**
 * @author THEMKR
 */
abstract class BaseFilter {

    val pixelArray: IntArray
    val imageWidth: Int
    val filterMatrixArrayList: ArrayList<FilterMatrix>

    /**
     * Constructor
     *
     * @param pixelArray            Array of Pixel
     * @param imageWidth
     * @param filterMatrixArrayList
     */
    constructor(pixelArray: IntArray, imageWidth: Int, filterMatrixArrayList: ArrayList<FilterMatrix>) {
        this.pixelArray = pixelArray
        this.imageWidth = imageWidth
        this.filterMatrixArrayList = filterMatrixArrayList
    }

    /**
     * Method to apply Filter and Get final Value
     *
     * @return Filtered PixelArray
     */
    @Throws(Exception::class)
    abstract fun applyFilter(): IntArray
}
